### ISSUE TITLE
An authentication issue with embedded problems in 2.19

### ISSUE THREAD

**Issue by lahvak** [openwebwork/webwork2#2548]
I have a website (generated using quarto) that includes exercises embedded in iframes.  I followed the example on this post:

https://michaelgage.blogspot.com/2015/06/whether-writing-full-text-book-or-just.html

An example:

```
<iframe title="Exercise" src="https://webwork.svsu.edu/webwork2/html2xml?
&amp;answersSubmitted=0
&amp;sourceFilePath=Library/272/setStewart12_1/problem_1.pg&amp;problemSeed=123
&amp;displayMode=MathJax
&amp;courseID=daemon_course
&amp;userID=daemon
&amp;course_password=daemon
&amp;outputformat=simple" width="640" height="500" frameborder="0"></iframe>
```

It worked fine in 2.18.  In 2.19, however, the problem initially renders correctly, but when trying to check answers, I get an "Authentication failed. Log in again to continue." message.

I also tried to convert the link to the new `render_rpc` endpoint:

```
<iframe title="Exercise" src="https://webwork.svsu.edu/webwork2/render_rpc?
&amp;answersSubmitted=0
&amp;sourceFilePath=Library/272/setStewart12_1/problem_1.pg&amp;problemSeed=123
&amp;displayMode=MathJax
&amp;courseID=daemon_course
&amp;user=daemon
&amp;passwd=daemon
&amp;outputformat=simple" width="640" height="500" frameborder="0"></iframe>
```

with the same result.  Initially the problem renders correctly, but after submission I get the "Authentication failed..." message.

The page is at https://stuff.lahvak.space/calc3/notes

**Comment by Alex-Jordan** [comment]
Try changing all the `&amp;`s to just `&`. I am not having trouble checking answers with:

https://webwork.svsu.edu/webwork2/render_rpc?&answersSubmitted=0&sourceFilePath=Library/272/setStewart12_1/problem_1.pg&problemSeed=123&displayMode=MathJax&courseID=daemon_course&user=daemon&passwd=daemon&outputformat=simple

**Comment by lahvak** [comment]
This is strange.

First, as far as I know, it should not make any difference whether plain & or &amp; is used.  I use the entity because that is whet quarto produces, so far I was not able to figure out how to prevent it from doing so.

Second, the link you gave works for me in chromium, but not in firefox.  In firefox it does the same as I described in the original issue: the problem renders correctly, but I get the authentication issue when I submit the answers.

--
Jan Hlavacek
Department of Mathematical Sciences, SVSU
(989) 964-2004 - https://space.lahvak.space/
________________________________

**Comment by Alex-Jordan** [comment]
There is some subtle thing that I can only vaguely recall about how hypnotoad versus apache2 handle the `&amp;`. Did you change to hypnotoad when upgrading to 2.19? 

For me, both Chrome and Firefox work with that URL and checking answers. But changing the `&` to `&amp;`, I get the authentication issue on the first page load, even before checking an answer.

**Comment by lahvak** [comment]
The entity (&amp; ) version will not work at all when typed in the address bar.  When linked from an html file, they both seem to work the same, though, which is what I would expect.

On the other hand, they both fail the same way (initial render is fine, but an authentication error on submit) when included in an iframe.  I may have to look in more details at the settings in webwork2.mojolicious.yml.

--
Jan Hlavacek
Department of Mathematical Sciences, SVSU
(989) 964-2004 - https://space.lahvak.space/
________________________________

**Comment by drgrice1** [comment]
Note that `&amp;` in a URL should not work as a URL parameter.  No browser supports using html entitites in a URL to denote a URL parameter.  The only reason it would work is if something translates that to a bare `&` before it actually goes to the browser.  When a URL is embedded in a page, whether or not it gets translated depends on the usage and browsers might do different things with html entities in links.  If the system you are using to generate html (quarto?) is changing `&` into `&amp;` that is a problem.  You might investigate if it has a way to properly insert links without them being escaped in that way.

This is not a valid issue for webwork2.  This is in reality an issue with the system that you are using to create these links, and there is nothing that can be done with the webwork2 code to fix this.  Furthermore, this has nothing to do with anything that was changed from changing webwork2 from mod_perl2 to Mojolicious.

**Comment by lahvak** [comment]
First, many documents on the web seem to disagree with you in the sense that in a HTML file, ampersand everywhere, including html tags attributes, should be entered as &amp;  You are correct that in the url sent to the server, the ampersands should appear as &, but that would not be valid in a html document (it seems to be more relaxed in html5, however, it still seems valid to use &amp; in such situation). The thing that translates &amp; into & is the browser, when it renders the HTML document into its internal representation.

It seems that in practice, browsers typically accept & in a html attribute as long as there is no ambiguity, but at least in html4, it seems that it would actually be invalid.

If you have a reference to an official document that says otherwise, it would be very useful, as I could point it out to quarto developers, but I could not find one.

It is however not relevant to this problem, since both of the following links will be accepted by WeBWorK:

https://webwork.svsu.edu/webwork2/render_rpc?&amp;answersSubmitted=0&amp;sourceFilePath=Library/272/setStewart12_1/problem_1.pg&amp;problemSeed=123&amp;displayMode=MathJax&amp;courseID=daemon_course&amp;user=daemon&amp;passwd=daemon &amp;outputformat=simple

and

https://webwork.svsu.edu/webwork2/render_rpc?&answersSubmitted=0&sourceFilePath=Library/272/setStewart12_1/problem_1.pg&problemSeed=123&displayMode=MathJax&courseID=daemon_course&user=daemon&passwd=daemon&outputformat=simple

when in an href attribute in an html document.

None of them will work for me when used in an iframe, both exhibiting the same authentication error when trying to submit the answer. The problem seems to be with the iframe, however, it is not a simple cross-origin issue (I did set the headers in webwork2.mojolicious.yml), since if that was the issue, I believe the problem would completely fail to render.

--
Jan Hlavacek
Department of Mathematical Sciences, SVSU
(989) 964-2004 - https://space.lahvak.space/
________________________________

**Comment by lahvak** [comment]
Note that if you click on the first of my two links in my previous comment, it will not work.  However, if you look at the source of this very page, you will see that the url in the source is encoded with `&amp;amp;`.  The browser renders the `&amp;` as &, and sends `&amp;` to the server, which is wrong.

The second link is encoded in the source with all the &'s replaced with `&amp;` even though I pasted it with &'s only. Similarly, the link that @Alex-Jordan writes in the comment https://github.com/openwebwork/webwork2/issues/2548#issuecomment-2320197686 as not having trouble with is encoded in the source using `&amp;` for each &.  If Quarto is incorrect about this, then so is github, but either way, the links with `&amp;` seem to work just fine.

**Comment by drgrice1** [comment]
When a link is used by the browser, it will not work with `&amp;` for the separator of a URL parameter.  If you are inserting a link into html, then depending on how that is inserted, it may need to be URL escaped.  If it is directly the content of the `href` attribute of an `a` tag, it does not.  If it is directly the content of the `src` attribute of an `iframe` tag, it does not.  However, if it is one of those things, but inside the content of a `srcdoc` attribute of an iframe tag, then it does need to be URL encoded.  In HTML5 it is not necessary (or even correct) to URL encode the `&` in an `href`, but most browsers will accept it and take care of URL decoding it appropriately.

You claim that your first link is accepted by webwork2 when it is the `href` attribute of an `a` tag, but that is incorrect.  What happens is that is URL decoded before the browser actually uses it as a link.  So when webwork2 gets it, it does not have the `&amp;` in it anymore.

In short, when the `&` should be URL encoded or not depends on the nature of how you are inserting it into the HTML code, and in the end it will only work properly as a link if something URL decodes it if needed so that when it is used as a link, it does not have `&amp;`.

This issue was closed because this is not a webwork2 bug regardless.  The issue is with your website itself, and how it handles the links that you are giving it.  The links in your page are not parsed by webwork2 at the time that they are inserted into your page, and there is nothing webwork2 can do about them.

**Comment by lahvak** [comment]
> When a link is used by the browser, it will not work with `&amp;` for the separator of a URL parameter. If you are inserting a link into html, then depending on how that is inserted, it may need to be URL escaped. If it is directly the content of the `href` attribute of an `a` tag, it does not. If it is directly the content of the `src` attribute of an `iframe` tag, it does not. However, if it is one of those things, but inside the content of a `srcdoc` attribute of an iframe tag, then it does need to be URL encoded. In HTML5 it is not necessary (or even correct) to URL encode the `&` in an `href`, but most browsers will accept it and take care of URL decoding it appropriately.
> 

I think I must not be communicating clearly, because there clearly is some misunderstanding.

> You claim that your first link is accepted by webwork2 when it is the `href` attribute of an `a` tag, but that is incorrect. What happens is that is URL decoded before the browser actually uses it as a link. So when webwork2 gets it, it does not have the `&amp;` in it anymore.
> 

But that's exactly what I am trying to say! In a html document, the link can (and some say should) encode ampersand as `&amp;`.  the browser change it to a simple ampersand before sending it to WeBWorK.  And that is exactly what happens with the contents generated by quarto.  Quarto (actually I believe it is done internally by pandoc) converts all ampersands in the link to `&amp;`.  The browser then converts them back to simple ampersands before sending the data to the server.

> This issue was closed because this is not a webwork2 bug regardless. The issue is with your website itself, and how it handles the links that you are giving it. The links in your page are not parsed by webwork2 at the time that they are inserted into your page, and there is nothing webwork2 can do about them.

As I already commented several times, it does not seem to be an issue with the ampersands.  Regardless whether the ampersands are encoded or not, the authentication issue persists.  I do believe it is probably an issue with our WeBWorK configuration, but I would like to at least figure out what I am doing wrong.  I think I followed the instructions in `webwork2.mojolicious.yml`, but still have the issue.  I would like to fix that.

Thank you.

**Comment by drgrice1** [comment]
If the ampersands are present in the link for the credentials provided in the link, then that would cause the issue with authentication.

I don't know who is telling you that you should URL encode the ampersands if used for an href attribute.  That is not correct.

In `webwork2.mojolicious.yml` have you set `allow_unsecured_rpc: 0`?  Also, if you are using the `render_rpc` endpoint, then you need to add `disableCookies=1` to the URL paramaters.

**Comment by Alex-Jordan** [comment]
I had a similar issue where the actual HTML source was `&amp;amp;` and that worked just fine when the web server was apache2, back when my WW server was 2.17. When I upgraded to 2.18, we switched to hypnotoad, and then these things stopped working. It took a minute to understand that the source had `&amp;amp;` because of course in many places, you will be shown `&amp;`. But making changes so that the source was just `&amp;` resolved things for me. This is what I was vaguely remembering about some difference with apache2. For whatever reason, it was able to handle `&amp;amp;` in URLs (even though it seems silly that it does that).

**Comment by lahvak** [comment]
`disableCookies=1` was the fix.  For some reason from the wording in the `webwork2.mojolicious.yml`, I got an impression that it is only needed when using 2FA, and since I don't use it for the daemon course, I did not think I needed it, and then I forgot about it.

I am sorry for wasting your time!
